### PR TITLE
Make route application dependent

### LINKED ISSUE #256: Make route application dependent
Under some precise circonstances, it may be useful to force a route to resolve itself on a given application which may be different from the current application that's running.

The idea is to make a route selfaware of the application it belongs so that it gets generated for that specific application regardless of the running application.

### PR DESCRIPTION
**Pull request by amazeika**
fixes #256

### CODE REVIEW

**Comment by amazeika** [comment]
@johanjanssens There's still something I would to fix on it, discussing it with @ercanozkaya, but all that we've discussed via email is there.

When running application and route application are the same, I bypass completely the code making the switch to improve BC, meaning that a regular JRoute::_ is made to avoid this from breaking if API changes.

File: code/libraries/joomlatools/component/koowa/dispatcher/router/route.php [hunk 1]
```diff
@@ -15,6 +15,54 @@
...
+        parent::__construct($config);
+
+        $this->_applications = KObjectConfig::unbox($config->applications);
+        $this->_application  = $this->setApplication($config->application);
+    }
+
+    protected function _initialize(KObjectConfig $config)
+    {
+        $config->append(array(
+            'applications' => array('site', 'administrator'),
```
Comment (johanjanssens) [inline code/libraries/joomlatools/component/koowa/dispatcher/router/route.php]: @amazeika Is there no api in Joomla to get these? Technically you can define your own in Joomla, site and administrator are the defaults provided.

Comment (amazeika) [reply code/libraries/joomlatools/component/koowa/dispatcher/router/route.php]: @johanjanssens The closest thing I've found to that is this https://github.com/joomla/joomla-cms/blob/staging/libraries/src/Application/ApplicationHelper.php#L121 ... not really sure what is a client in this context though ... not very clear.

Comment (johanjanssens) [reply code/libraries/joomlatools/component/koowa/dispatcher/router/route.php]: @amazeika That is the correct API. A client is the same as an application. I called it client, in hindsight 12 years later application has been the better name.

Comment (amazeika) [reply code/libraries/joomlatools/component/koowa/dispatcher/router/route.php]: @johanjanssens alright, I'll use that then :)

File: code/libraries/joomlatools/component/koowa/dispatcher/router/route.php [hunk 2]
```diff
@@ -39,8 +39,16 @@ public function __construct(KObjectConfig $config)
 
     protected function _initialize(KObjectConfig $config)
     {
+        $clients = JApplicationHelper::getClientInfo();
```
Comment (amazeika) [inline code/libraries/joomlatools/component/koowa/dispatcher/router/route.php]: @johanjanssens Added this here as requested

**Pull request merged**